File: LeetCode/289. Game of Life.py
# https://leetcode.com/problems/game-of-life/description/

from test import test


class Solution:
    def gameOfLife(self, board: list[list[int]]) -> list[list[int]]:
        max_x = len(board[0]) - 1
        max_y = len(board) - 1

        def should_change_state(x, y):
            is_alive = bool(board[y][x])
            alive = 0

            for y_coord in range(max(0, y - 1), min(max_y, y + 1) + 1):
                for x_coord in range(max(0, x - 1), min(max_x, x + 1) + 1):
                    if x_coord == x and y_coord == y:
                        continue

                    if board[y_coord][x_coord]:
                        alive += 1

            if not is_alive:
                return alive == 3

            return alive < 2 or alive > 3

        change_state: list[tuple[int, int]] = []

        for y in range(max_y + 1):
            for x in range(max_x + 1):
                if should_change_state(x, y):
                    change_state.append((x, y))

        for coords in change_state:
            x, y = coords
            print(x, y)

            board[y][x] = 0 if board[y][x] else 1

        return board


solution = Solution()


test(
    solution.gameOfLife,
    [
        {
            "input": [
                [
                    [0, 1, 0],
                    [0, 0, 1],
                    [1, 1, 1],
                    [0, 0, 0],
                ],
            ],
            "expected": [
                [0, 0, 0],
                [1, 0, 1],
                [0, 1, 1],
                [0, 1, 0],
            ],
        },
        {
            "input": [
                [
                    [1, 1],
                    [1, 0],
                ],
            ],
            "expected": [
                [1, 1],
                [1, 1],
            ],
        },
    ],
)

File: LeetCode/289. Game of Life.js
// https://leetcode.com/problems/game-of-life/description/

const { test } = require('../test');

const gameOfLife = (board) => {
  const xMax = board[0].length - 1;
  const yMax = board.length - 1;

  const shouldChangeState = (x, y) => {
    const isAlive = board[y][x];

    let alive = 0;

    for (let i = Math.max(y - 1, 0); i <= Math.min(y + 1, yMax); i++) {
      for (let k = Math.max(x - 1, 0); k <= Math.min(x + 1, xMax); k++) {
        if (i === y && k === x) continue;

        board[i][k] && alive++;
      }
    }

    if (!isAlive) {
      return alive === 3;
    }

    return alive < 2 || alive > 3;
  };

  const changeState = [];

  for (let y = 0; y <= yMax; y++) {
    for (let x = 0; x <= xMax; x++) {
      if (shouldChangeState(x, y)) {
        changeState.push([x, y]);
      }
    }
  }

  for (const coords of changeState) {
    const [x, y] = coords;

    board[y][x] = board[y][x] ? 0 : 1;
  }

  return board;
};

test(gameOfLife, [
  {
    input: [
      [
        [0, 1, 0],
        [0, 0, 1],
        [1, 1, 1],
        [0, 0, 0],
      ],
    ],
    expected: [
      [
        [0, 0, 0],
        [1, 0, 1],
        [0, 1, 1],
        [0, 1, 0],
      ],
    ],
  },
  {
    input: [
      [
        [1, 1],
        [1, 0],
      ],
    ],
    expected: [
      [1, 1],
      [1, 1],
    ],
  },
]);
